src/transformers/models/git/modeling_git.py

```diff
@@ -0,0 +1,1399 @@
+        r"""
+        labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
+            Labels for computing the left-to-right language modeling loss (next word prediction). Indices should be in
+            `[-100, 0, ..., config.vocab_size]` (see `input_ids` docstring) Tokens with indices set to `-100` are
+            ignored (masked), the loss is only computed for the tokens with labels n `[0, ..., config.vocab_size]`
+        past_key_values (`tuple(tuple(torch.FloatTensor))` of length `config.n_layers` with each tuple having 4 tensors of shape `(batch_size, num_heads, sequence_length - 1, embed_size_per_head)`):
+            Contains precomputed key and value hidden states of the attention blocks. Can be used to speed up decoding.
+
+            If `past_key_values` are used, the user can optionally input only the last `decoder_input_ids` (those that
+            don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of all
+            `decoder_input_ids` of shape `(batch_size, sequence_length)`.
+        use_cache (`bool`, *optional*):
+            If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding (see
+            `past_key_values`).
+        """
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+        if labels is not None:
+            use_cache = False
+
+        outputs = self.git(
+            input_ids,
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            pixel_values=pixel_values,
+            head_mask=head_mask,
+            inputs_embeds=inputs_embeds,
+            past_key_values=past_key_values,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        sequence_output = outputs[0]
+        logits = self.output(sequence_output)
+
+        lm_loss = None
+        if labels is not None:
+            # we are doing next-token prediction; shift prediction scores and input ids by one
+            shifted_logits = logits[:, :-1, :].contiguous()
+            labels = labels[:, 1:].contiguous()
+            loss_fct = CrossEntropyLoss()
+            lm_loss = loss_fct(shifted_logits.view(-1, self.config.vocab_size), labels.view(-1))
+
+        if not return_dict:
+            output = (logits,) + outputs[1:]
+            return ((lm_loss,) + output) if lm_loss is not None else output
+
+        return CausalLMOutputWithPast(
+            loss=lm_loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
+
+    def prepare_inputs_for_generation(self, input_ids, past=None, attention_mask=None, use_cache=True, **model_kwargs):
```

Comment:
Full context for generate integration:

The output of this function is passed to the model forward pass (see [here](https://github.com/huggingface/transformers/blob/4bb07647504a277398856e828fa48ddbec97678e/src/transformers/generation/utils.py#L1934) and in the lines below).

For encoder-decoder models, the input to the model is handled before it gets here, and this function just needs to pull `encoder_outputs` from `model_kwargs` and place it in the output (see [Bart](https://github.com/huggingface/transformers/blob/4bb07647504a277398856e828fa48ddbec97678e/src/transformers/models/bart/modeling_bart.py#L1411) for an example)

For Decoder-only models, it is slightly more complex. First, `generate` assumes that the model's `self.main_input_name` will be passed as its first argument OR with a matching argument (i.e. `model.generate(pixel_values)` and `model.generate(pixel_values=pixel_values)` will work, `model.generate(inputs=pixel_values)` won't). All other model inputs should be passed as keyword arguments (e.g. `model.generate(foo=foo)`), and they should be an input to `prepare_inputs_for_generation` -- `**model_kwargs` confirms that all arguments are used. `prepare_inputs_for_generation` should return all the inputs the model needs.

From what I wrote above, it seems like this model can be directly used in `generate`, with one caveat: at the moment, we throw an exception for decoder-only models when `self.main_input_name` is not `input_ids` ([here](https://github.com/huggingface/transformers/blob/4bb07647504a277398856e828fa48ddbec97678e/src/transformers/generation/utils.py#L520)). We can remove this check, it is redundant with the more complete set of checks that I've added recently :)

Reply:
For this particular model, main_input_name is still input_ids. There's just an additional input for each time step, namely pixel_values.

I tried to do `model.generate(input_ids, model_kwargs={"pixel_values":pixel_values}, max_length=20)` as well as adding "pixel_values" as key to the output of prepare_inputs_for_generation, but no luck so far.